### DISCUSSION TITLE
Provider Schemes

### DISCUSSION THREAD

**Discussion by nikpivkin** [opentofu/opentofu#1004] [General]
Hi All. The [tofu providers schema](https://opentofu.org/docs/cli/commands/providers/schema) command prints the provider schema. But is there any way to get provider schema using Go and `opentofu` package? . Almost all `opentofu` code is in the internal package, which cannot be imported.

**Comment by ghost** [comment]
Hi @nikpivkin the short-term answer is unfortunately no, if I'm not mistaken. However, I had a chat with @Yantrio this morning about this and it would be cool to expose the data from OpenTofu in a library format.

**Comment by apparentlymart** [comment]
A while back I started implementing a library for interacting with providers from other Go programs outside of OpenTofu, by reimplementing the same protocol as a client.

I lost interest before I fully finished it, but I _did_ get far enough to write [an example that extracts schema information from a provider](https://github.com/apparentlymart/terraform-provider/blob/edf80d31c6b51cceeccdc783cdc9859bea970438/examples/tfplugin-example-types/main.go). The schema data retrieved by that example program is the same data that `tofu providers schema` starts from, so you should be able to use that to get equivalent information as long as it's something I'd got around to building support for.

I'm not intending to work on it any further in the near future, but it's permissively licensed in case you'd like to fork it and extend it with support for any other parts of the schema I didn't implement yet.

(This is a personal project of mine, _not_ an OpenTofu project.)

**Reply by nikpivkin** [reply]
Hi, thanks!
